src/lib/storage/frame_of_reference/frame_of_reference_encoder.hpp

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include <array>
+#include <limits>
+#include <memory>
+
+#include "storage/base_column_encoder.hpp"
+
+#include "storage/frame_of_reference_column.hpp"
+#include "storage/value_column.hpp"
+#include "storage/value_column/value_column_iterable.hpp"
+#include "storage/vector_compression/vector_compression.hpp"
+#include "types.hpp"
+#include "utils/enum_constant.hpp"
+
+namespace opossum {
+
+class FrameOfReferenceEncoder : public ColumnEncoder<FrameOfReferenceEncoder> {
+ public:
+  static constexpr auto _encoding_type = enum_c<EncodingType, EncodingType::FrameOfReference>;
+  static constexpr auto _uses_vector_compression = true;
+
+  template <typename T>
+  std::shared_ptr<BaseEncodedColumn> _on_encode(const std::shared_ptr<const ValueColumn<T>>& value_column) {
+    const auto alloc = value_column->values().get_allocator();
+
+    static constexpr auto block_size = FrameOfReferenceColumn<T>::block_size;
+
+    const auto size = value_column->size();
+    const auto num_blocks = size / block_size;
+
+    auto block_minima = pmr_vector<T>{alloc};
+    block_minima.reserve(num_blocks);
+
+    auto offset_values = pmr_vector<uint32_t>{alloc};
+    offset_values.reserve(size);
+
+    auto null_values = pmr_vector<bool>{alloc};
+    null_values.reserve(size);
+
+    auto max_value = uint32_t{0u};
+
+    auto iterable = ValueColumnIterable<T>{*value_column};
+    iterable.with_iterators([&](auto column_it, auto column_end) {
+      auto value_block = std::array<T, block_size>{};
+
+      while (column_it != column_end) {
+        auto value_block_it = value_block.begin();
+        for (; value_block_it != value_block.end() && column_it != column_end; ++value_block_it, ++column_it) {
```

Comment:
I originally wrote this:

> Maybe I am missing something, but I would expect you to create a new `value_block` (I would call it `current_value_block`) once it is full. I can‘t see this happening. Are you overwriting the one block by any chance?

After further reading I found that this is only a temp structure.  I think it would help if you added some comments to the five variables defined above (also: What is max_value used for?)

Reply:
`current_value_block` makes sense. `max_value` is passed into `compress_vector`. It is used for the fixed-size byte-aligned vector to determine its width. I can add some comments.